### PR TITLE
Feature/actual streaming

### PR DESCRIPTION
**Pull request by gsheasby**
The diff bottom line looks like more than it actually is - I introduce a new generated file, KeyValueTable, which takes over 700 of the 1100 lines.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1341)

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+import org.apache.commons.lang3.ArrayUtils;
+
+public final class BlockConsumingInputStream extends InputStream {
+    private final BlockGetter blockGetter;
+    private final long numBlocks;
+    private final int blocksInMemory;
+
+    private int nextBlockToRead;
+    private Iterator<Byte> buffer;
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java]: this scares me from performance perspective in both iterating over single bytes as well as boxing overhead. Do we have any numbers on throughput for this?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    private void getNextBlock() throws IOException {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        int numBlocksToGet = Math.min(blocksLeft(), blocksInMemory);
+        blockGetter.get(nextBlockToRead, numBlocksToGet, outputStream);
+        nextBlockToRead += numBlocksToGet;
+        byte[] bytes = outputStream.toByteArray();
+        List<Byte> list = Arrays.asList(ArrayUtils.toObject(bytes));
+        buffer = list.iterator();
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java]: I think the internal buffer here should be a `byte[]` not an `Iterator<Byte>`

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            return buffer.next() & 0xff;
+        }
+
+        return -1;
+    }
+
+    private void getNextBlock() throws IOException {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        int numBlocksToGet = Math.min(blocksLeft(), blocksInMemory);
+        blockGetter.get(nextBlockToRead, numBlocksToGet, outputStream);
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java]: how much does this actually read from the block getter and write to the output stream? can we presize output stream to avoid expansions?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockGetter.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.stream;
+
+import java.io.OutputStream;
+
+interface BlockGetter {
+    void get(Integer firstBlock, Integer numBlocks, OutputStream destination);
```
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockGetter.java]: They take boxed primitives as a hangover from how the code used to look like - it was a templated `BiConsumer<Integer, OutputStream>`, because you can't do `BiConsumer<int, OutputStream>`. Now that we have our own interface, I can change this back.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java [hunk 5]
```diff
@@ -150,47 +180,206 @@ private long storeAndCheckByteStreams(int size) throws IOException {
...
+        stream.close();
+    }
+
+    @Test
+    public void testOverwrite() throws IOException {
+        Random rand = new Random();
+        StreamTestTableFactory tableFactory = StreamTestTableFactory.of();
+
+        final byte[] reference = PtBytes.toBytes("ref");
+        final byte[] bytes1 = new byte[2 * StreamTestStreamStore.BLOCK_SIZE_IN_BYTES];
```
Comment (schlosna) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java]: we need some fencepost tests around the `BLOCK_SIZE_IN_BYTES` and `IN_MEMORY_THRESHOLD`

**Comment by gsheasby** [comment]
There is an open PR (#1231) for stream benchmarking, but we deprioritised work on it.

I've written a quick and dirty test that I've used to confirm that, as @schlosna mentioned, throughput is adversely affected, and there are a few simple ways to fix them up. I'll do this.

For posterity, this is the test I've run with various sizes (100, 200, 500 MB)
```
    @Test
    public void perfTest() throws IOException {
        int megabytes = 200;
        final int size = megabytes * 1024 * 1024;
        byte[] bytesToStore = new byte[size];
        Random rand = new Random();
        rand.nextBytes(bytesToStore);

        final long id = storeStream(bytesToStore, PtBytes.toBytes(megabytes));

        Stopwatch timer = Stopwatch.createStarted();
        InputStream stream = txManager.runTaskThrowOnConflict(t -> store.loadStream(t, id));
        timer.stop();
        long millis = timer.elapsed(TimeUnit.MILLISECONDS);

        byte[] sample = new byte[size];
        Stopwatch loadTimer = Stopwatch.createStarted();
        int bytesRead = stream.read(sample);
        loadTimer.stop();
        long loadMillis = loadTimer.elapsed(TimeUnit.MILLISECONDS);
        System.out.println(String.format("Load stream for %d megabytes took %d milliseconds.", megabytes, millis));
        System.out.println(String.format("Actually stream for %d megabytes took %d milliseconds.", megabytes, loadMillis));
    }
```

**Comment by gsheasby** [comment]
The test above measured the "before" code at about 400 MB/s for loading the stream, and about 1250 MB/s for actually streaming the data (300 MB/s end to end).

The original PR loads the stream in O(1) time, but streamed the data very slowly (70-120MB/s). This has boosted to about 275MB/s with the latest commit.

Next, I'll override some the other `InputStream.read` methods, which I think will bring overall throughput above the 300MB/s mark we had before.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java [hunk 6]
```diff
@@ -101,11 +100,33 @@ private InputStream tryGetStream(Transaction transaction, ID id, StreamMetadata
...
         } else {
-            File file = loadToNewTempFile(transaction, id, metadata);
-            return new DeleteOnCloseFileInputStream(file);
+            return makeStream(id, metadata);
         }
     }
 
+    private InputStream makeStream(ID id, StreamMetadata metadata) {
+        BlockGetter pageRefresher =
+                (firstBlock, numBlocks, outputStream) ->
```
Comment (sandorw) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: numBlocks was initially a bit confusing here, since it's also used below as the total number of blocks, rather than the number to fetch and hold in memory with each refresh. I'm not sure of the behavior of opening/committing transactions inside other transactions,

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java (hunk 3 above)
Comment (sandorw) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java]: Alternately, we could read directly into the buffer and avoid the BAOS, although this complicates the BlockGetter interface.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java [hunk 7]
```diff
@@ -101,11 +100,33 @@ private InputStream tryGetStream(Transaction transaction, ID id, StreamMetadata
             loadSingleBlockToOutputStream(transaction, id, 0, ios);
             return ios.getInputStream();
         } else {
-            File file = loadToNewTempFile(transaction, id, metadata);
-            return new DeleteOnCloseFileInputStream(file);
+            return makeStream(id, metadata);
```
Comment (sandorw) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: The first stream block loaded is from a new transaction, whereas this method already receives an open transaction. Is this intended?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+        if (nextBlockToRead < numBlocks) {
+            getNextBlock();
+            return buffer[positionInBuffer++] & 0xff;
+        }
+
+        return -1;
+    }
+
+    private void getNextBlock() throws IOException {
```
Comment (sandorw) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java]: nit: potentially loading many stream blocks

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java (hunk 3 above)
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java]: That depends on the `BlockGetter` implementation. In this case, we call `StreamStore.loadBlocksToOutputStream`.

To allow us to predict the size of the output stream, I have added a `getExpectedSize` method to `BlockGetter`.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java (hunk 7 above)
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: Yes. The original transaction is assumed to have been closed by the time we get to loading the second block.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            blockGetter.get(nextBlockToRead, numBlocksToGet, outputStream);
+            nextBlockToRead += numBlocksToGet;
+            buffer = outputStream.toByteArray();
+            positionInBuffer = 0;
+            return true;
+        }
+    }
+
+    private int blocksLeft() {
+        return numBlocks - nextBlockToRead;
```
Comment (SerialVelocity) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java]: This should probably be `Math.max(0,`.

File: atlasdb-client/src/test/java/com/palantir/atlasdb/stream/BlockConsumingInputStreamTest.java [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    private void expectNextBytesFromStream(BlockConsumingInputStream stream, String expectedOutput) throws IOException {
+        byte[] chunk = new byte[2];
+        int read = stream.read(chunk);
+        assertEquals(2, read);
+        assertArrayEquals(expectedOutput.getBytes(), chunk);
+    }
+
+    private class MockableBlockGetter implements BlockGetter {
```
Comment (SerialVelocity) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/stream/BlockConsumingInputStreamTest.java]: Why was this necessary? You should be able to spy on the `singleByteGetter` directly.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java [hunk 11]
```diff
@@ -150,47 +185,206 @@ private long storeAndCheckByteStreams(int size) throws IOException {
...
-        verifyLoadingStreams(id, bytesToStore, store);
+        verifyLoadingStreams(id, bytesToStore);
+
+        return id;
+    }
+
+    @Test
+    public void shouldQuicklyGetFirstByte() throws IOException {
+        long millisFor100mb = measureMillisToGetFirstByteFromStreamStoreWithMbs(100);
+        long millisFor1mb = measureMillisToGetFirstByteFromStreamStoreWithMbs(1);
```
Comment (SerialVelocity) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java]: We should put these in benchmarking if anywhere. Circle seems to have random spikes every so often.

File: docs/source/schemas/streams.rst [hunk 12]
```diff
@@ -40,6 +44,23 @@ this is to write a ``KeyValueService`` implementation that just
...
 one and other tables to another.
 
+Transactionality
+================
+
+When you call ``loadStream(transaction, id)``, only the first section of the stream data is pre-loaded as part of that transaction.
+The amount loaded is determined by the block size (1MB) and the in-memory threshold (4MiB by default); we load at least one block,
+a whole number of blocks, and (if the in-memory threshold is at least the block size) as many blocks as fit inside the in-memory threshold.
+So by default, we load 4 blocks (``4*10^6`` bytes, slightly less than the default in-memory threshold of ``4*2^20`` bytes).
+If your stream does not inside the in-memory threshold, then the remainder of the data will be buffered in separate transactions.
```
Comment (SerialVelocity) [inline docs/source/schemas/streams.rst]: `does not fit inside`

File: docs/source/schemas/streams.rst [hunk 13]
```diff
@@ -40,6 +44,23 @@ this is to write a ``KeyValueService`` implementation that just
...
+
+This has some subtle transactionality implications.
+Suppose you store a mapping of keys to stream IDs, and at ``start_timestamp``, ``key`` maps to ``stream_id_1``.
+You then call ``loadStream(transaction,stream_id_1)``, and keep the resulting ``InputStream`` open.
+Before you reach the end of the stream, another thread stores a mapping from ``key`` to ``stream_id_2``.
+In this case, streaming of the original stream will continue and complete successfully, even though the data being streamed is out of date.
```
Comment (SerialVelocity) [inline docs/source/schemas/streams.rst]: Would this have failed before? Can the old stream be deleted because of the reference count reaching 0?

File: atlasdb-client/src/test/java/com/palantir/atlasdb/stream/BlockConsumingInputStreamTest.java [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+            }
+        }
+
+        @Override
+        public int expectedLength() {
+            return data.length;
+        }
+    };
+
+    private final byte[] stored = "divisible".getBytes();
```
Comment (SerialVelocity) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/stream/BlockConsumingInputStreamTest.java]: These should all have the charset listed (UTF-8)

**Comment by sandorw** [comment]
Just noticed that the PR referenced in the release notes is not this PR.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java [hunk 15]
```diff
@@ -101,11 +100,43 @@ private InputStream tryGetStream(Transaction transaction, ID id, StreamMetadata
...
+        try {
+            return BlockConsumingInputStream.create(pageRefresher, numBlocks, blocksInMemory);
+        } catch (IOException e) {
+            throw Throwables.throwUncheckedException(e);
+        }
+    }
+
+    private int numberOfBlocksThatFitInMemory() {
+        long blocksInMemory = getInMemoryThreshold() / BLOCK_SIZE_IN_BYTES;
+        return Math.max(1, (int) blocksInMemory);
```
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: We bow down at the feet of the machine which can hold over 2 petabytes in memory.

Ben suggested using long everywhere, which would eliminate this problem.

Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: The smallest thing we can load is a block. If `blocksInMemory` is set to 0, then `BlockConsumingInputStream.getNextBlock` would never load a block.

File: atlasdb-client/src/test/java/com/palantir/atlasdb/stream/BlockConsumingInputStreamTest.java (hunk 10 above)
Comment (gsheasby) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/stream/BlockConsumingInputStreamTest.java]: Aha. It was necessary when `singleByteGetter` was only a lambda.  Fixed.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java (hunk 11 above)
Comment (gsheasby) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java]: We're rather permissive here, but I get your point. There's a couple of tests I could put in #1231.

File: docs/source/schemas/streams.rst (hunk 13 above)
Comment (gsheasby) [inline docs/source/schemas/streams.rst]: No - we were never actually straming the data before - we saved the stream to a temporary file, and then streamed from that.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java (hunk 15 above)
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: On closer inspection, I note that `buffer.length` is an int, so the true maximum number of blocks is `Integer.MAX_VALUE / BLOCK_SIZE_IN_BYTES = 2147`. It's worth adding a test for that, as it's only about 2GB.

**Comment by gsheasby** [comment]
This, from [build 5352](https://circleci.com/gh/palantir/atlasdb/5352?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link), makes me sad :(

shouldQuicklyGetFirstByte - com.palantir.atlasdb.schema.stream.StreamTest
morejava.lang.AssertionError: Loading the first byte of a stream should perform better than log(size). We actually expect constant time, but are being permissive. Fetching 1MB took 2ms, but 100 MB took 664ms.

**Comment by schlosna** [comment]
@gsheasby do you know what happened with that build 5352? Can we assert that we only read/fetch one block in each case?

**Comment by gsheasby** [comment]
@schlosna It looks to have been a circle flake - the problem with using circle to measure perf like that is, it's highly susceptible to random noise (e.g. other builds using the same underlying box). The perf-testing project is a much better fit for this sort of test.

**Comment by schlosna** [comment]
Yeah, rather than relying on the timing which may flake, was thinking we could spy on the block getter to verify it is only called once in both cases.

File: docs/source/schemas/streams.rst (hunk 13 above)
Comment (nziebart) [inline docs/source/schemas/streams.rst]: Hmm, trying to understand the issue here. Since `stream_id` corresponds to exactly one immutable stream, I don't see why your transactional guarantees are any worse here than if you stored the entire stream in the `key` row. It seems like you could consider `stream_id` as a compressed `stream`, and consider reading the `stream` as decompressing it.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java [hunk 16]
```diff
@@ -101,11 +100,44 @@ private InputStream tryGetStream(Transaction transaction, ID id, StreamMetadata
...
 
+    private InputStream makeStream(ID id, StreamMetadata metadata) {
+        long totalBlocks = getNumberOfBlocksFromMetadata(metadata);
+        int blocksInMemory = getNumberOfBlocksThatFitInMemory();
+
+        BlockGetter pageRefresher = new BlockGetter() {
+            @Override
+            public void get(long firstBlock, long numBlocks, OutputStream destination) {
+                txnMgr.runTaskReadOnly(txn -> {
+                    for (int i = 0; i < numBlocks; i++) {
```
Comment (SerialVelocity) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: Should probably be `long` just in case.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+            return buffer[positionInBuffer++] & 0xff;
+        }
+
+        return -1;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        Preconditions.checkNotNull(b, "Cannot read into a null array!");
+        if (off < 0 || len < 0 || len > b.length - off) {
```
Comment (SerialVelocity) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java]: I think `len > b.length - off` is a valid case. `len` is the maximum amount of bytes to return so if you hit the end of the stream before that, you just return less bytes.

Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/BlockConsumingInputStream.java]: That check comes from the javadoc of InputStream, which I'd rather not contradict.

**Comment by gsheasby** [comment]
@schlosna The test `canLoadMultipleBlocksAtOnceAndAlsoFewerBlocksAtEnd` in `BlockConsumingInputStreamTest` does this.

**Comment by gsheasby** [comment]
Solved #1150.
